### PR TITLE
Support decoding pa_data as part of kdc enc response

### PR DESCRIPTION
**Pull request by adfoster-r7**
Fixes a crash when the AS_REP response contains pa_data when decoding the enc_kdc_response:

```
[-] 192.168.123.136:445 - Exploit failed: Rex::Proto::Kerberos::Model::Error::KerberosDecodingError Failed to decode tag 12 in ENC-KDC-RESPONSE SEQUENCE
```

## Verification

- [ ] Ensure CI passes

### CODE REVIEW

File: spec/lib/rex/proto/kerberos/model/enc_kdc_response_spec.rb [hunk 1]
```diff
@@ -304,6 +345,18 @@
...
       end
+
+      it "decodes the pa_data correctly" do
+        enc_kdc_response.decode(enc_as_resp)
+        expected_pa_data = [
+          Rex::Proto::Kerberos::Model::PreAuthData.new(
+            type: Rex::Proto::Kerberos::Model::PA_SUPPORTED_ETYPES,
+            # SupportedEnctypes: 0x0000001f, des-cbc-crc, des-cbc-md5, rc4-hmac, aes128-cts-hmac-sha1-96, aes256-cts-hmac-sha1-96
+            value: "\x1F\x00\x00\x00"
+          )
```
Comment (adfoster-r7) [inline spec/lib/rex/proto/kerberos/model/enc_kdc_response_spec.rb:356]: It looks like there's a few dedicated pa type objects in framework for request scenarios, but they aren't wired up to the decoding scenarios:

- pre_auth_enc_time_stamp.rb
- pre_auth_pac_request.rb

I've opted to leave that out for now.

File: spec/lib/rex/proto/kerberos/model/enc_kdc_response_spec.rb [hunk 2]
```diff
@@ -225,10 +225,49 @@
...
+                @tag=2,
+                @tag_class=:CONTEXT_SPECIFIC,
+                @value=
+                 [#<OpenSSL::ASN1::OctetString:0x00007ffd80e8cc60
+                   @indefinite_length=false,
+                   @tag=4,
+                   @tag_class=:UNIVERSAL,
+                   @tagging=nil,
+                   @value="\x1F\x00\x00\x00">]>]>]>]>]>]>
 =end
```
Comment (adfoster-r7) [inline spec/lib/rex/proto/kerberos/model/enc_kdc_response_spec.rb:268]: I'm not sure of the use of this debug output on all of the existing Kerberos tests, but I've kept it for consistency.

File: lib/rex/proto/kerberos/model/enc_kdc_response.rb [hunk 3]
```diff
@@ -4,6 +4,22 @@ module Rex
...
+        #            nonce           [2] UInt32,
+        #            key-expiration  [3] KerberosTime OPTIONAL,
+        #            flags           [4] TicketFlags,
+        #            authtime        [5] KerberosTime,
+        #            starttime       [6] KerberosTime OPTIONAL,
+        #            endtime         [7] KerberosTime,
+        #            renew-till      [8] KerberosTime OPTIONAL,
+        #            srealm          [9] Realm,
+        #            sname           [10] PrincipalName,
+        #            caddr           [11] HostAddresses OPTIONAL
```
Comment (adfoster-r7) [inline lib/rex/proto/kerberos/model/enc_kdc_response.rb:20]: This isn't decoded yet, I've created a separate ticket for this. I haven't seen the KDC return this value.

File: lib/rex/proto/kerberos/model/element.rb [hunk 4]
```diff
@@ -30,6 +30,8 @@ def self.decode(input)
           end
 
           def initialize(options = {})
+            raise ArgumentError, "Invalid options, expected hash, got #{options.class}" unless options.is_a?(Hash)
```
Comment (adfoster-r7) [inline lib/rex/proto/kerberos/model/element.rb:33]: Added this to let folk know you can't do `KerberosObject.new(asn1)`. Happy to revert.

**Review by dwelch-r7 — APPROVED** [review]

**Comment by dwelch-r7** [comment]
# Release Notes

Support decoding Privileged Attribute Data as part of kdc encrypted response

**Pull request merged**